src/aws_iot_mqtt_client.c

```diff
@@ -169,6 +169,39 @@ IoT_Error_t aws_iot_mqtt_set_connect_params(AWS_IoT_Client *pClient, IoT_Client_
 	FUNC_EXIT_RC(SUCCESS);
 }
 
+IoT_Error_t aws_iot_mqtt_free(AWS_IoT_Client *pClient)
+{
+    IoT_Error_t rc = SUCCESS;
+
+    if (NULL == pClient) {
+        rc = NULL_VALUE_ERROR;
+    }else
+	{
+	#ifdef _ENABLE_THREAD_SUPPORT_
+		if (rc == SUCCESS)
+		{
+			rc = aws_iot_thread_mutex_destroy(&(pClient->clientData.state_change_mutex));
+		}
+
+		if (rc == SUCCESS)
+		{
+			rc = aws_iot_thread_mutex_destroy(&(pClient->clientData.tls_read_mutex));
+		}else{
+			(void)aws_iot_thread_mutex_destroy(&(pClient->clientData.tls_read_mutex));
+		}
+
+		if (rc == SUCCESS)
+		{
+			rc = aws_iot_thread_mutex_destroy(&(pClient->clientData.tls_write_mutex));
+		}else{
+			(void)aws_iot_thread_mutex_destroy(&(pClient->clientData.tls_read_mutex));
```

Comment:
This line should destroy `tls_write_mutex`.

Also, is it necessary to re-assign `rc`?  Seems we could just do the following to clean everything up.
```c
if(rc == SUCCESS) /* Check not NULL_VALUE_ERROR. */
{
    (void)aws_iot_thread_mutex_destroy(&(pClient->clientData.state_change_mutex));
    (void)aws_iot_thread_mutex_destroy(&(pClient->clientData.tls_read_mutex));
    (void)aws_iot_thread_mutex_destroy(&(pClient->clientData.tls_write_mutex));
}
```

Reply:
I though of that but it will not work unless you refactor the function because it has exit points everywhere